rllib/agents/impala/vtrace_torch_policy.py

```diff
@@ -251,51 +163,198 @@ def make_time_major(policy, seq_lens, tensor, drop_last=False):
+        else:
+            return torch.optim.RMSprop(
+                params=self.model.parameters(),
+                lr=self.cur_lr,
+                weight_decay=self.config["decay"],
+                momentum=self.config["momentum"],
+                eps=self.config["epsilon"],
+            )
+
+
+# VTrace mixins are placed in front of more general mixins to make sure
+# their functions like optimizer() overrides all the other implementations
+# (e.g., LearningRateSchedule.optimizer())
+class VTraceTorchPolicy(
+    GradClippingMixin,
+    VTraceOptimizer,
+    LearningRateSchedule,
+    EntropyCoeffSchedule,
+    TorchPolicyV2,
+):
+    """PyTorch policy class used with PPOTrainer."""
+
+    def __init__(self, observation_space, action_space, config):
+        config = dict(ray.rllib.agents.impala.impala.DEFAULT_CONFIG, **config)
+
+        GradClippingMixin.__init__(self)
+        VTraceOptimizer.__init__(self)
+        # Need to initialize learning rate variable before calling
+        # TorchPolicyV2.__init__.
+        LearningRateSchedule.__init__(self, config["lr"], config["lr_schedule"])
+        EntropyCoeffSchedule.__init__(
+            self, config["entropy_coeff"], config["entropy_coeff_schedule"]
+        )
 
-def choose_optimizer(policy, config):
-    if policy.config["opt_type"] == "adam":
-        return torch.optim.Adam(params=policy.model.parameters(), lr=policy.cur_lr)
-    else:
-        return torch.optim.RMSprop(
-            params=policy.model.parameters(),
-            lr=policy.cur_lr,
-            weight_decay=config["decay"],
-            momentum=config["momentum"],
-            eps=config["epsilon"],
+        TorchPolicyV2.__init__(
+            self,
+            observation_space,
+            action_space,
+            config,
+            max_seq_len=config["model"]["max_seq_len"],
         )
 
+        # TODO: Don't require users to call this manually.
+        self._initialize_loss_from_dummy_batch()
 
-def setup_mixins(policy, obs_space, action_space, config):
-    EntropyCoeffSchedule.__init__(
-        policy, config["entropy_coeff"], config["entropy_coeff_schedule"]
-    )
-    LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
-
-
-VTraceTorchPolicy = build_policy_class(
-    name="VTraceTorchPolicy",
-    framework="torch",
-    loss_fn=build_vtrace_loss,
-    get_default_config=lambda: ray.rllib.agents.impala.impala.DEFAULT_CONFIG,
-    stats_fn=stats,
-    extra_grad_process_fn=apply_grad_clipping,
-    optimizer_fn=choose_optimizer,
-    before_init=setup_mixins,
-    mixins=[LearningRateSchedule, EntropyCoeffSchedule],
-    get_batch_divisibility_req=lambda p: p.config["rollout_fragment_length"],
-)
+    @override(TorchPolicyV2)
+    def loss(
+        self,
+        model: ModelV2,
+        dist_class: Type[ActionDistribution],
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
+        """Compute loss for Impala Policy Objective.
```

Comment:
Actually, do we even need these docstrings? We are overriding a method from a super class. Normally, we don't repeat dosctrings in these cases.

Reply:
yeah, I don't think we need them. ok, getting rid of as much as I can find.